### PR TITLE
add new tutorial questdb and tensorflow

### PR DESCRIPTION
**Pull request by pswu11**

### CODE REVIEW

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 1]
```diff
@@ -0,0 +1,540 @@
...
+
+### Importing Your Dependencies
+
+Now that your dependencies are installed, it’s time to start implementing the
+time series forecasting with TensorFlow and QuestDB. If you want to clone the
+project and follow along in your own [Jupyter](https://jupyter.org/) notebook,
+[here is the link](https://github.com/gouravsinghbais/Time-Series-Forecasting-with-Tensorflow-and-QuestDB)
+to the GitHub repo. First, import the following dependencies:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:208]: There should be `python` after the three backticks.
Same for the rest of the examples here in this page.

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 2]
```diff
@@ -0,0 +1,540 @@
...
+### Installing QuestDB
+
+To install QuestDB on any OS platform, you’ll need to:
+
+1. [Install Docker](https://docs.docker.com/engine/install/)
+
+2. Pull the QuestDB Docker image and create a Docker container. Open your
+   command prompt and write the following command:
+
+   ```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:183]: shell

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+Let’s get started!
+
+### Installing TensorFlow
+
+TensorFlow can be easily installed with a Python package manager (PIP). If you
+have [Anaconda](https://www.anaconda.com/) installed in your system, you can use
+the Anaconda prompt. For normal Python installation, you can use the default
+command prompt and write the following command to install TensorFlow:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:167]: shell

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 4]
```diff
@@ -0,0 +1,540 @@
...
+large amounts of data. Thus, you’ll use QuestDB to store the time series data
+instead. To do so, first make sure the QuestDB Docker container is running and
+accessible:
+
+```
+## create table query
+q = 'create table excel_rates'\
+	'(Date timestamp,'\
+	'USD int,'\
+	'INR double)'
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:269]: More readable Python is:

```python
q = '''create table excel_rates (
           Date timestamp,
           USD int,
           INR double)'''
```

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 5]
```diff
@@ -0,0 +1,540 @@
...
+fail = 0
+
+## iterate over each row and store it in the QuestDB table
+for i, row in df.iterrows():
+	date = row['Date']
+	## convert date to datetime format to store in DB
+	date = "'"+date.strftime('%Y-%m-%dT%H:%M:%S.%fZ')+"'"
+	usd = row['USD']
+	inr = row['INR']
+	query = "insert into excel_rates values(" + date + "," + str(usd) + ","  + str(inr) +")"
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:301]: Line too long, also, not very Pythonic. Suggesting:

```python
query = f'insert into excel_rates values({date}, {usd}, {inr})'
```

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 6]
```diff
@@ -0,0 +1,540 @@
...
+accessible:
+
+```
+## create table query
+q = 'create table excel_rates'\
+	'(Date timestamp,'\
+	'USD int,'\
+	'INR double)'
+## connect to QuestDB URL and execute the query
+r = requests.get("http://localhost:9000/exec?query=" + q)
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:271]: Dodgy code, query not escaped. Should read:

```python
r = requests.get('http://localhost:9000/exec', params={'query': q})
```

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 7]
```diff
@@ -0,0 +1,540 @@
...
+
+## iterate over each row and store it in the QuestDB table
+for i, row in df.iterrows():
+	date = row['Date']
+	## convert date to datetime format to store in DB
+	date = "'"+date.strftime('%Y-%m-%dT%H:%M:%S.%fZ')+"'"
+	usd = row['USD']
+	inr = row['INR']
+	query = "insert into excel_rates values(" + date + "," + str(usd) + ","  + str(inr) +")"
+	r = requests.get("http://localhost:9000/exec?query=" + query)
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:302]:
```python
r = requests.get('http://localhost:9000/exec', params={'query': query})
```

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 8]
```diff
@@ -0,0 +1,540 @@
...
+
+### Selecting the Data from QuestDB
+
+Now your data is stored in the QuestDB time series database. However, storing
+the data is not enough; you need to be able to access it in order to do the time
+series forecasting. Use this code to access the data in QuestDB:
+
+```
+## select data from QuestDB
+r = requests.get("http://localhost:9000/exp?query=select * from excel_rates")
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:331]:
```python
r = requests.get('http://localhost:9000/exec', params={'query': 'select * from excel_rates'})
```

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 9]
```diff
@@ -0,0 +1,540 @@
...
+The graphs displaying actual labels and predicted labels should look like this:
+
+![Model testing](/img/tutorial/2022-03-30/model-testing-1.png)
+
+As you can see, the predictions are the approximation of actual values, which
+indicates that the model is performing well enough. Since the testing data has a
+lot of points, the graph appears clustered; to take a closer look at the values,
+you can visualize just one hundred values as well:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:506]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 10]
```diff
@@ -0,0 +1,540 @@
...
+
+Once you execute the above code, your model training will start. It should look
+something like this:
+
+![Model training](/img/tutorial/2022-03-30/model-training.png)
+
+Once the model is ready, you’ll need to test it on the test set to check its
+accuracy:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:486]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 11]
```diff
@@ -0,0 +1,540 @@
...
+y_test = test_scaled[:,-1]
+X_test = test_scaled[:,0:-1]
+```
+
+### Creating the Model
+
+Now, the preprocessing is complete, and it’s time to train your deep learning
+(GRU) model for time series forecasting. Use this code to do so:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:454]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 12]
```diff
@@ -0,0 +1,540 @@
...
+data = np.array(data)
+## you can take last 500 data points as test set
+train , test = data[0:-500], data[-500:]
+```
+
+Since there is diversity in the data (values are varying at a large scale), you
+need to apply some scaling or normalization to the data. For this purpose, you
+will be using **MinMaxScaler**:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:433]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 13]
```diff
@@ -0,0 +1,540 @@
...
+data.head()
+```
+
+Once complete, your data set should look like this:
+
+![Preprocessed data](/img/tutorial/2022-03-30/processed-data.png)
+
+Next, you need to split the data into two different categories—train and test:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:422]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 14]
```diff
@@ -0,0 +1,540 @@
...
+## remove 0 values
+indexed_df = indexed_df[indexed_df.INR != 0.0]
+```
+
+After removing the '0' values, you will have to flip the data frame to sort it
+from 2009 to 2022:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:393]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 15]
```diff
@@ -0,0 +1,540 @@
...
+The plot should look something like this:
+
+![All data set values](/img/tutorial/2022-03-30/plot-all-values.png)
+
+As you can see, the values prior to 2009 all appear as “0.” Because this can
+affect the accuracy of the model, go ahead and remove them. Thankfully, there
+are plenty of non-zero values after 2009 included in the data set. Use the
+following code to remove the '0' values:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:385]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 16]
```diff
@@ -0,0 +1,540 @@
...
+
+You should now see the DataFrame as follows:
+
+![Data](/img/tutorial/2022-03-30/data.png)
+
+To see how the value of INR is varying according to time, you can plot a curve
+between time and INR using this code:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:371]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 17]
```diff
@@ -0,0 +1,540 @@
...
+
+At this point in the tutorial, you’ve created the table, inserted data into the
+table, and read the data from the table. Now, it’s time to do some time series
+forecasting preprocessing.
+
+First, you can go ahead and remove the USD column, as it contains all '1'
+values, and as such, it will not contribute to the forecasting in any way. To do
+so, use the following code:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:354]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 18]
```diff
@@ -0,0 +1,540 @@
...
+QuestDB is running (in this case, port 9000) and execute the query using the
+request module of Python. If the query execution is successful, it will return
+the status code **200**; if not, you will receive the status code **400**.
+
+### Inserting the Data to QuestDB
+
+Once you have created the table, you need to store your data in it. Use the
+following code to do so:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:289]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 19]
```diff
@@ -0,0 +1,540 @@
...
+![Read Dataset](/img/tutorial/2022-03-30/read-dataset.png)
+
+### Creating QuestDB Tables
+
+Now you’ve seen your data set, but Excel is limited in its ability to store
+large amounts of data. Thus, you’ll use QuestDB to store the time series data
+instead. To do so, first make sure the QuestDB Docker container is running and
+accessible:
+
+```
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:264]: python (see earlier comment)

File: tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md [hunk 20]
```diff
@@ -0,0 +1,540 @@
...
+  alt="QuestDB log and Tensorflow logo"
+  height={467}
+  src="/img/tutorial/2022-03-30/banner.png"
+  width={650}
+></Banner>
+
+This post comes from
+[Gourav Singh Bais](https://www.linkedin.com/in/gourav-singh-bais/), who has
+written an excellent tutorial that shows how to build an application that uses
+time series data to forecast trends and events using Tensorflow and QuestDB.
```
Comment (amunra) [inline tutorial/2022-03-30-forecasting-with-questdb-tensorflow.md:25]: We tend to say "timeseries" instead of "time series" throughout the rest of our documents.
I think this (and all other instances) need updating.

**Review by amunra — CHANGES_REQUESTED** [review]
